src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -733,162 +765,215 @@ DatabaseShardImp::onChildrenStopped()
 void
 DatabaseShardImp::import(Database& source)
 {
+    assert(init_);
+
+    // Only the application local node store can be imported
+    if (&source != &app_.getNodeStore())
     {
-        std::lock_guard lock(mutex_);
-        assert(init_);
+        assert(false);
+        JLOG(j_.error()) << "invalid source database";
+        return;
+    }
+
+    std::uint32_t earliestIndex;
+    std::uint32_t latestIndex;
+    {
+        auto loadLedger = [&](bool ascendSort =
+                                  true) -> boost::optional<std::uint32_t> {
+            std::shared_ptr<Ledger> ledger;
+            std::uint32_t ledgerSeq;
+            std::tie(ledger, ledgerSeq, std::ignore) = loadLedgerHelper(
+                "WHERE LedgerSeq >= " + std::to_string(earliestLedgerSeq()) +
+                    " order by LedgerSeq " + (ascendSort ? "asc" : "desc") +
+                    " limit 1",
+                app_,
+                false);
+            if (!ledger || ledgerSeq == 0)
+            {
+                JLOG(j_.error()) << "no suitable ledgers were found in"
+                                    " the SQLite database to import";
+                return boost::none;
+            }
+            return ledgerSeq;
+        };
+
+        // Find earliest ledger sequence stored
+        auto ledgerSeq{loadLedger()};
+        if (!ledgerSeq)
+            return;
+        earliestIndex = seqToShardIndex(*ledgerSeq);
 
-        // Only the application local node store can be imported
-        if (&source != &app_.getNodeStore())
+        // Consider only complete shards
+        if (ledgerSeq != firstLedgerSeq(earliestIndex))
+            ++earliestIndex;
+
+        // Find last ledger sequence stored
+        ledgerSeq = loadLedger(false);
+        if (!ledgerSeq)
+            return;
+        latestIndex = seqToShardIndex(*ledgerSeq);
+
+        // Consider only complete shards
+        if (ledgerSeq != lastLedgerSeq(latestIndex))
+            --latestIndex;
+
+        if (latestIndex < earliestIndex)
         {
-            assert(false);
-            JLOG(j_.error()) << "invalid source database";
+            JLOG(j_.error()) << "no suitable ledgers were found in"
+                                " the SQLite database to import";
             return;
         }
+    }
+
+    auto numHistShards = numHistoricalShards(std::lock_guard(mutex_));
+    auto const sweepInterval = std::chrono::milliseconds(500);
+
+    // Sweep the source and family caches to conserve
+    // memory
+    ThreadWrapper mainSweeper(
+        [&source, sweepInterval, &family = app_.getNodeFamily()] {
+            source.sweep();
+            family.sweep();
+            std::this_thread::sleep_for(sweepInterval);
+        });
 
-        std::uint32_t earliestIndex;
-        std::uint32_t latestIndex;
+    std::cout << std::endl
```

Comment:
I agree it's unusual. I chose to write directly to `cout` because in order for the progress bars to be displayed correctly they can't be interlaced with any other output. The easiest way to achieve this was to suppress logging to the console with the `--silent` argument. The caveat however, as you've noticed, is that the code then breaks a convention of sorts by writing to `cout` directly. If you think this design choice is egregious, I can look into alternative approaches.

Reply:
Nah, I think the pattern is fine.  But you need to keep in mind the folks who will stumble through here three years from now and be inclined to "fix the bug".  Leave a paragraph comment here in the source code (where maintainers are likely to find it) and explain what's up.

Reply:
Done